app/uk/gov/hmrc/emcstfereferencedata/repositories/ExciseProductsRepository.scala

```diff
@@ -81,4 +83,31 @@ class ExciseProductsRepository @Inject() (val mongoComponent: MongoComponent)(us
       .find()
       .sort(ascending("code"))
       .toFuture()
+
+  def fetchProductCodesInformation(
+    cnInformationRequest: CnInformationRequest
+  ): Future[Map[String, CnCodeInformation]] = {
+
+    val productCodeToCnCode: Map[String, String] =
+      cnInformationRequest.items.map(item => item.productCode -> item.cnCode).toMap
+
+    collection
+      .find(in("code", cnInformationRequest.items.map(_.productCode)*))
+      .sort(ascending("code"))
+      .toFuture()
+      .map { products =>
+        products.flatMap { product =>
+          productCodeToCnCode.get(product.code).map { cnCode =>
+            val cnInfo = CnCodeInformation(
+              cnCode = cnCode,
+              cnCodeDescription = product.description,
+              exciseProductCode = product.code,
+              exciseProductCodeDescription = product.description,
+              unitOfMeasureCode = product.unitOfMeasureCode
+            )
+            cnCode -> cnInfo
+          }
+        }.toMap
```

Comment:
```suggestion
        val productsByCode = products.map(product => product.code -> product).toMap
        items.flatMap { item =>
          productsByCode.get(item.productCode).map { product =>
            val cnInfo = CnCodeInformation(
              cnCode = item.cnCode,
              cnCodeDescription = product.description,
              exciseProductCode = product.code,
              exciseProductCodeDescription = product.description,
              unitOfMeasureCode = product.unitOfMeasureCode
            )
            item.cnCode -> cnInfo
          }
        }.toMap
```
Maybe we could do this mapping in the other direction instead?